src/main/java/io/anserini/collection/C4Collection.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.anserini.collection;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+import java.util.zip.GZIPInputStream;
+
+
+public class C4Collection extends DocumentCollection<C4Collection.Document> {
+    public C4Collection(Path path) {
+        this.path = path;
+    }
+
+    @Override
+    public FileSegment<C4Collection.Document> createFileSegment(Path p) throws IOException {
+        return new Segment(p);
+    }
+
+    public static class Segment extends FileSegment<C4Collection.Document>{
+        private MappingIterator<JsonNode> iterator; // iterator for JSON line objects
+        private JsonNode node = null;
+        private String fileName;
+        private int count = 0;
+
+        public Segment(Path path) throws IOException {
+            super(path);
+            fileName = path.toString();
+            if (fileName.endsWith(".gz")) { //.gz
+                InputStream stream = new GZIPInputStream(
+                        Files.newInputStream(path, StandardOpenOption.READ), BUFFER_SIZE);
+                bufferedReader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
+            } else { // plain text file
+                bufferedReader = new BufferedReader(new FileReader(fileName));
+            }
+            // reading as a json file
+            ObjectMapper mapper = new ObjectMapper();
+            // removes ctrl characters
+            String filtered = bufferedReader.lines().map(line -> line.replaceAll("[\\p{C}]","")).collect(Collectors.joining());
+            iterator = mapper.readerFor(JsonNode.class).readValues(filtered);
+            if(iterator.hasNext()){
+                node = iterator.next();
+            }
+        }
+
+        @Override
+        public void readNext() throws NoSuchElementException {
+            if (node == null) {
+                throw new NoSuchElementException("JsonNode is empty");
+            } else {
+                bufferedRecord = new C4Collection.Document(node, fileName, count);
+                if (iterator.hasNext()) { // if bufferedReader contains JSON line objects, we parse the next JSON into node
```

Comment:
I don't think you need this logic here because C4 doesn't change formats?

Reply:
don't we need this for when we reach eof though?

Reply:
The comment is a bit confusing, I think? The entire file is JSON line, right?